main/eval/src/mill/eval/EvaluatorCore.scala

```diff
@@ -66,93 +66,117 @@ private[mill] trait EvaluatorCore extends GroupEvaluator {
+      // necessary upstream futures will have already been scheduled and stored,
+      // due to the topological order of traversal.
+      for (terminal <- terminals) {
+        val deps = interGroupDeps(terminal)
+        futures(terminal) = Future.sequence(deps.map(futures)).map { upstreamValues =>
+          if (failed.get()) None
+          else {
+            val upstreamResults = upstreamValues
+              .iterator
+              .flatMap(_.iterator.flatMap(_.newResults))
+              .toMap
+
+            val startTime = System.nanoTime() / 1000
+            val threadId = threadNumberer.getThreadId(Thread.currentThread())
+            val counterMsg = s"${count.getAndIncrement()}/${terminals.size}"
+            val contextLogger = PrefixLogger(
+              out = logger,
+              context = contextLoggerMsg(threadId),
+              tickerContext = GroupEvaluator.dynamicTickerPrefix.value
+            )
+
+            val res = evaluateGroupCached(
+              terminal = terminal,
+              group = sortedGroups.lookupKey(terminal),
+              results = upstreamResults,
+              counterMsg = counterMsg,
+              zincProblemReporter = reporter,
+              testReporter = testReporter,
+              logger = contextLogger
             )
-          )
 
-          Some(res)
+            if (failFast && res.newResults.values.exists(_.result.asSuccess.isEmpty))
+              failed.set(true)
+
+            val endTime = System.nanoTime() / 1000
+
+            val duration = endTime - startTime
+
+            chromeProfileLogger.log(
+              task = Terminal.printTerm(terminal),
+              cat = "job",
+              startTime = startTime,
+              duration = duration,
+              threadId = threadNumberer.getThreadId(Thread.currentThread()),
+              cached = res.cached
+            )
+
+            profileLogger.log(
+              ProfileLogger.Timing(
+                terminal.render,
+                (duration / 1000).toInt,
+                res.cached,
+                deps.map(_.render),
+                res.inputsHash,
+                res.previousInputsHash
+              )
+            )
+
+            Some(res)
+          }
         }
       }
     }
 
-    val finishedOptsMap = terminals
+    val tasks0 = terminals0.filter {
+      case Terminal.Labelled(c: Command[_], _) => false
+      case _ => true
+    }
+
+    val (_, tasksTransitive0) = Plan.plan(Agg.from(tasks0.map(_.task)))
+
+    val tasksTransitive = tasksTransitive0.toSet
+    val (tasks, commands) = terminals0.partition {
+      case Terminal.Labelled(t, _) if tasksTransitive.contains(t) => true
+      case _ => false
+    }
+
+    // Run all non-command tasks according to the threads
+    // given but run the commands in linear order
+    evaluateTerminals(
+      tasks,
+      if (sys.env.contains("MILL_TEST_SUITE")) _ => ""
```

Comment:
Can you add a comment with some explanation here?

I guess, time is ripe to have a table with all used environment variables in our readme.

Reply:
Done. This is an internal environment variable, so users should not need to know about it

Reply:
I had other developers (myself included) in mind. I thought, our readme is targeting devlopers.